app/models/katello/host/content_facet.rb

```diff
@@ -164,10 +164,12 @@ def update_repositories_by_paths(paths)
         prefixes = %w(/pulp/deb/ /pulp/repos/ /pulp/content/)
         relative_paths = []
 
-        # paths == ["/pulp/content/Default_Organization/Library/custom/Test_product/test2"]
+        # paths == ["/pulp/content/Default_Organization/Library/custom/Test_product/test2",
+        #           "/pulp/content/Org/Library/custom/Test_product/test2/%3Fcomp%3Dmain%26rel%3Dstable"]
```

Comment:
are you sure that its test2**/**%3F...
I thought its test2?comp ...

Reply:
On our test-system subman had it encoded in the baseurl and therefore also reported it that way to katello.
We may have to investigate, if this is an issue itself :thinking:

Reply:
The weird characters making it into the path is a Candlepin bug, I don't have a link to it but someone should have it somewhere.

Reply:
It was fixed a while back I believe.

Reply:
https://github.com/candlepin/candlepin/pull/4783

Reply:
It is definitely not fixed here http://yum.theforeman.org/candlepin/4.4/.

Also, I still think this particular issue should really be fixed on the client side. It also is an issue independently of the encoding issue, because the mere fact that there are URL params (encoded or not) breaks applicability, because Katello does not recognize the repos.

Reply:
The fix should be in Candlepin 4.4.12-1 or later. Also you would need to regenerate your entitlement certificates (subscription-manager identity --regenerate --force).

thanks to @evgeni for info

Reply:
Our test system is using `candlepin-4.4.14-1.el8.noarch` and I am seeing subman output with `%3F`:
```
# subscription-manager repos
+----------------------------------------------------------+
    Available Repositories in /etc/yum.repos.d/redhat.repo
+----------------------------------------------------------+
Repo ID:   ATIX_Saltstack_for_Debian_12_Saltstack_for_Debian_12
Repo Name: Saltstack for Debian 12
Repo URL:  https://orcharhino.testdmz.atix/pulp/deb/ATIX/Testing/Composite_Debian_12/custom/Saltstack_for_Debian_12/Saltstack_for_Debian_12/%3Fcomp%3Dmain%26rel%3Dbookworm
Enabled:   0

Repo ID:   ATIX_Debian_12_bookworm-updates_main
Repo Name: bookworm-updates main
Repo URL:  https://orcharhino.testdmz.atix/pulp/deb/ATIX/Testing/Composite_Debian_12/custom/Debian_12/bookworm-updates_main/%3Fcomp%3Dmain%26rel%3Dbookworm-updates
Enabled:   1
```
Running `subscription-manager identity --regenerate --force` has no effect on this.

Is it possible that the bug that was fixed for entitlement certificates was just adjacent to what we are using for structured APT?

Either way, we have made the implementation robust both for encoded and not encoded URL options. :shrug:

Reply:
Also, I am not sure if I made this properly clear, we need the extra processing here regardless of whether candlepin passes strings with `%3F` or strings with `?`. Either way, subman is reporting a list of repositories to Katello with structured APT options appended to the relative path bit. Katello then compares this list to it's repositories without the structured APT options appended, and so finds no matches. In both cases we need to strip away the structured APT options before the comparison. Which is what we are doing several lines down.

Without this fix Katello never shows any applicable or upgradeable packages, because it does not know what repos to look in.

Reply:
I've gotten confirmation from Candlepin team that the URL encoding in your example is now expected:

"Yeah, the changes URL encode the segments, so I would expect that you'd see artifacts of that with the underscore -> %3F example."

So I think we're good here, as long as this code works the way you want it to.